presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java

```diff
@@ -0,0 +1,575 @@
+
+        remainder = (getInt(decimal, 0) & LONG_MASK) + (remainder << 32);
+        int z0 = (int) (remainder / divisor);
+
+        pack(result, z0, z1, z2, z3, isNegative(decimal));
+    }
+
+    private static void setNegative(Slice decimal, boolean negative)
+    {
+        setNegative(decimal, getInt(decimal, SIGN_INT_INDEX), negative);
+    }
+
+    private static void setNegative(Slice decimal, int v3, boolean negative)
+    {
+        setRawInt(decimal, SIGN_INT_INDEX, v3 | (negative ? SIGN_INT_MASK : 0));
+    }
+
+    private static void pack(Slice decimal, int v0, int v1, int v2, int v3, boolean negative)
+    {
+        setRawInt(decimal, 0, v0);
+        setRawInt(decimal, 1, v1);
+        setRawInt(decimal, 2, v2);
+        setNegative(decimal, v3, negative);
+    }
+
+    private static void throwIfOverflows(int v0, int v1, int v2, int v3)
+    {
+        if (exceedsOrEqualTenToThirtyEight(v0, v1, v2, v3)) {
+            throwOverflowException();
+        }
+    }
+
+    private static void throwIfOverflows(Slice decimal)
+    {
+        if (exceedsOrEqualTenToThirtyEight(decimal)) {
+            throwOverflowException();
+        }
+    }
+
+    private static boolean exceedsOrEqualTenToThirtyEight(int v0, int v1, int v2, int v3)
+    {
+        // 10**38=
+        // v0=0(0),v1=160047680(98a22400),v2=1518781562(5a86c47a),v3=1262177448(4b3b4ca8)
+
+        if (v3 >= 0 && v3 < 0x4b3b4ca8) {
+            return false;
+        }
+        else if (v3 > 0x4b3b4ca8) {
+            return true;
+        }
+
+        if (v2 >= 0 && v2 < 0x5a86c47a) {
+            return false;
+        }
+        else if (v2 > 0x5a86c47a) {
+            return true;
+        }
+
+        return v1 < 0 || v1 >= 0x098a2240;
+    }
+
+    private static boolean exceedsOrEqualTenToThirtyEight(Slice decimal)
+    {
+        // 10**38=
+        // v0=0(0),v1=160047680(98a22400),v2=1518781562(5a86c47a),v3=1262177448(4b3b4ca8)
+
+        int v3 = getInt(decimal, 3);
+        if (v3 >= 0 && v3 < 0x4b3b4ca8) {
+            return false;
+        }
+        else if (v3 > 0x4b3b4ca8) {
+            return true;
+        }
+
+        int v2 = getInt(decimal, 2);
+        if (v2 >= 0 && v2 < 0x5a86c47a) {
+            return false;
+        }
+        else if (v2 > 0x5a86c47a) {
+            return true;
+        }
+
+        int v1 = getInt(decimal, 1);
+        return v1 < 0 || v1 >= 0x098a2240;
+    }
+
+    private static void throwOverflowException()
+    {
+        throw new ArithmeticException("Decimal overflow");
+    }
+
+    /* Based on: it.unimi.dsi.fastutil.bytes.ByteArrays.reverse; */
+    private static byte[] reverse(final byte[] a)
+    {
+        final int length = a.length;
+        for (int i = length / 2; i-- != 0; ) {
+            final byte t = a[length - i - 1];
+            a[length - i - 1] = a[i];
+            a[i] = t;
+        }
+
+        return a;
+    }
+
+    private static int getInt(Slice decimal, int index)
```

Comment:
there is `getRawInt` if you want sign flag too. JIT should remove branches anyway because index is constant value